File: packages/shared/src/rbac.ts
/**
 * Role-based access control matrix.
 *
 * ENFORCEMENT LIVES IN RUST. Every Tauri IPC command calls
 * `pos_core::rbac::authorize` before touching data or hardware. This TypeScript
 * copy exists only so the UI can hide controls a role cannot use — hiding a
 * button is never a security boundary.
 *
 * Both copies are pinned to `contracts/rbac.json`; a drift fails CI.
 */
import { z } from 'zod';

export const ROLES = ['owner', 'manager', 'cashier'] as const;
export const RoleSchema = z.enum(ROLES);
export type Role = z.infer<typeof RoleSchema>;

export const PERMISSIONS = [
  // Sales floor
  'catalog.view',
  'customer.lookup',
  'sale.create',
  'receipt.print',
  'drawer.kick',
  'loyalty.redeem',
  // Supervisor
  'sale.refund',
  'sale.void',
  'discount.apply',
  'shift.open',
  'shift.close',
  'receipt.reprint',
  'report.view',
  'report.z_run',
  'inventory.view',
  // Back office
  'catalog.manage',
  'inventory.adjust',
  'customer.manage',
  'supplier.manage',
  'purchase_order.manage',
  'analytics.view',
  'audit.view',
  'user.manage',
  'settings.manage',
] as const;
export const PermissionSchema = z.enum(PERMISSIONS);
export type Permission = z.infer<typeof PermissionSchema>;

const CASHIER: readonly Permission[] = [
  'catalog.view',
  'customer.lookup',
  'sale.create',
  'receipt.print',
  'drawer.kick',
  'loyalty.redeem',
];

const MANAGER: readonly Permission[] = [
  ...CASHIER,
  'sale.refund',
  'sale.void',
  'discount.apply',
  'shift.open',
  'shift.close',
  'receipt.reprint',
  'report.view',
  'report.z_run',
  'inventory.view',
];

export const ROLE_PERMISSIONS: Readonly<Record<Role, ReadonlySet<Permission>>> = {
  owner: new Set(PERMISSIONS),
  manager: new Set(MANAGER),
  cashier: new Set(CASHIER),
};

export function hasPermission(role: Role, permission: Permission): boolean {
  return ROLE_PERMISSIONS[role].has(permission);
}

/** Login PIN: 4–6 digits, entered on the on-screen numpad. */
export const PinSchema = z.string().regex(/^\d{4,6}$/, 'PIN must be 4–6 digits');

File: crates/pos-core/src/rbac.rs
//! Role-based access control — the enforcement point.
//!
//! Every IPC command that reads or mutates business data calls [`authorize`]
//! with the session's role *before* doing anything else. The TypeScript copy
//! of this matrix only drives what the UI shows.

use serde::{Deserialize, Serialize};

use crate::error::{IpcError, IpcResult};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum Role {
    Owner,
    Manager,
    Cashier,
}

impl Role {
    pub const ALL: [Role; 3] = [Role::Owner, Role::Manager, Role::Cashier];

    pub fn as_str(self) -> &'static str {
        match self {
            Role::Owner => "owner",
            Role::Manager => "manager",
            Role::Cashier => "cashier",
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
pub enum Permission {
    // Sales floor
    #[serde(rename = "catalog.view")]
    CatalogView,
    #[serde(rename = "customer.lookup")]
    CustomerLookup,
    #[serde(rename = "sale.create")]
    SaleCreate,
    #[serde(rename = "receipt.print")]
    ReceiptPrint,
    #[serde(rename = "drawer.kick")]
    DrawerKick,
    #[serde(rename = "loyalty.redeem")]
    LoyaltyRedeem,
    // Supervisor
    #[serde(rename = "sale.refund")]
    SaleRefund,
    #[serde(rename = "sale.void")]
    SaleVoid,
    #[serde(rename = "discount.apply")]
    DiscountApply,
    #[serde(rename = "shift.open")]
    ShiftOpen,
    #[serde(rename = "shift.close")]
    ShiftClose,
    #[serde(rename = "receipt.reprint")]
    ReceiptReprint,
    #[serde(rename = "report.view")]
    ReportView,
    #[serde(rename = "report.z_run")]
    ReportZRun,
    #[serde(rename = "inventory.view")]
    InventoryView,
    // Back office
    #[serde(rename = "catalog.manage")]
    CatalogManage,
    #[serde(rename = "inventory.adjust")]
    InventoryAdjust,
    #[serde(rename = "customer.manage")]
    CustomerManage,
    #[serde(rename = "supplier.manage")]
    SupplierManage,
    #[serde(rename = "purchase_order.manage")]
    PurchaseOrderManage,
    #[serde(rename = "analytics.view")]
    AnalyticsView,
    #[serde(rename = "audit.view")]
    AuditView,
    #[serde(rename = "user.manage")]
    UserManage,
    #[serde(rename = "settings.manage")]
    SettingsManage,
}

impl Permission {
    pub const ALL: [Permission; 24] = [
        Permission::CatalogView,
        Permission::CustomerLookup,
        Permission::SaleCreate,
        Permission::ReceiptPrint,
        Permission::DrawerKick,
        Permission::LoyaltyRedeem,
        Permission::SaleRefund,
        Permission::SaleVoid,
        Permission::DiscountApply,
        Permission::ShiftOpen,
        Permission::ShiftClose,
        Permission::ReceiptReprint,
        Permission::ReportView,
        Permission::ReportZRun,
        Permission::InventoryView,
        Permission::CatalogManage,
        Permission::InventoryAdjust,
        Permission::CustomerManage,
        Permission::SupplierManage,
        Permission::PurchaseOrderManage,
        Permission::AnalyticsView,
        Permission::AuditView,
        Permission::UserManage,
        Permission::SettingsManage,
    ];
}

/// The matrix. An exhaustive `match` — adding a permission without deciding
/// who gets it is a compile error.
pub const fn is_allowed(role: Role, permission: Permission) -> bool {
    use Permission::*;
    match permission {
        CatalogView | CustomerLookup | SaleCreate | ReceiptPrint | DrawerKick | LoyaltyRedeem => {
            true
        }
        SaleRefund | SaleVoid | DiscountApply | ShiftOpen | ShiftClose | ReceiptReprint
        | ReportView | ReportZRun | InventoryView => matches!(role, Role::Owner | Role::Manager),
        CatalogManage | InventoryAdjust | CustomerManage | SupplierManage | PurchaseOrderManage
        | AnalyticsView | AuditView | UserManage | SettingsManage => matches!(role, Role::Owner),
    }
}

/// Guard used at the top of every privileged IPC command.
pub fn authorize(role: Role, permission: Permission) -> IpcResult<()> {
    if is_allowed(role, permission) {
        Ok(())
    } else {
        Err(IpcError::forbidden(format!(
            "The {} role is not allowed to perform this action.",
            role.as_str()
        )))
    }
}

pub fn permissions_for(role: Role) -> Vec<Permission> {
    Permission::ALL
        .into_iter()
        .filter(|p| is_allowed(role, *p))
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn hierarchy_holds() {
        for p in Permission::ALL {
            if is_allowed(Role::Cashier, p) {
                assert!(is_allowed(Role::Manager, p), "{p:?}");
            }
            if is_allowed(Role::Manager, p) {
                assert!(is_allowed(Role::Owner, p), "{p:?}");
            }
            assert!(is_allowed(Role::Owner, p));
        }
    }

    #[test]
    fn authorize_rejects_with_forbidden() {
        let err = authorize(Role::Cashier, Permission::SaleRefund).expect_err("must be denied");
        assert_eq!(err.code, crate::IpcErrorCode::Forbidden);
        assert!(authorize(Role::Manager, Permission::SaleRefund).is_ok());
    }
}
